cpp/src/shim/libcurl.cpp

```diff
@@ -116,19 +119,55 @@ CURL* CurlHandle::handle() noexcept { return _handle.get(); }
 
 void CurlHandle::perform()
 {
-  // Perform the curl operation and check for errors.
-  CURLcode err = curl_easy_perform(handle());
-  if (err != CURLE_OK) {
-    std::string msg(_errbuf);  // We can do this because we always initialize `_errbuf` as empty.
-    std::stringstream ss;
-    ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
-    if (msg.empty()) {
-      ss << "(" << curl_easy_strerror(err) << ")";
+  long http_code          = 0;
+  int attempt_count       = 1;
+  int base_delay          = 500;   // milliseconds
+  int max_delay           = 4000;  // milliseconds
+  auto http_max_attempts  = kvikio::defaults::http_max_attempts();
+  auto& http_status_codes = kvikio::defaults::http_status_codes();
+
+  while (attempt_count <= http_max_attempts) {
+    auto err = curl_easy_perform(handle());
+    curl_easy_getinfo(handle(), CURLINFO_RESPONSE_CODE, &http_code);
+
+    // Check if we should retry based on HTTP status code
+    if (std::find(http_status_codes.begin(), http_status_codes.end(), http_code) !=
+        http_status_codes.end()) {
+      // Retry only if one of the specified status codes is returned
+      // TODO: Parse the Retry-After header, if it exists.
+      // TODO: configurable maximum wait.
+      if (attempt_count == http_max_attempts) {
+        std::stringstream ss;
+        ss << "kvikio http_max_attempts_reached. attempts=" << http_max_attempts
+           << " reason=" << http_code;
```

Comment:
I'd use something more human-readable, like this:
```suggestion
        ss << "KvikIO: HTTP request reached maximum number of attempts (" << http_max_attempts << "). Got HTTP code " << http_code << "."
```